### PR TITLE
Refactor transaction processing in banking stage

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Transaction processing is hard to follow in the banking stage

#### Summary of Changes
- Simplify the `record_transactions` method so that it's clearer that it's all or nothing
- Refactor how the `commit_transactions_result` list of booleans is created

Fixes #

### CODE REVIEW

File: core/src/banking_stage.rs [hunk 1]
```diff
@@ -1204,68 +1168,78 @@ impl BankingStage {
...
-        // mark transactions that were executed but not recorded
-        retryable_record_transaction_indexes.iter().for_each(|i| {
-            transactions_execute_and_record_status[*i] = false;
-        });
+        if let Err(recorder_err) = record_transactions_result {
+            let retryable_record_transaction_indexes: Vec<usize> = execution_results
+                .iter()
+                .enumerate()
+                .filter_map(|(index, execution_result)| {
+                    execution_result.was_executed().then(|| index)
```
Comment (carllin) [inline core/src/banking_stage.rs:1216]: could probably directly push into `retryable_transaction_indexes` and then for the 
```
inc_new_counter_info!(
 "banking_stage-record_transactions_retryable_record_txs",
retryable_record_transaction_indexes.len()
);
```

counter below, just increment by the `executed_transactions_count`

File: core/src/banking_stage.rs [hunk 2]
```diff
@@ -1315,9 +1295,9 @@ impl BankingStage {
                 "find_and_send_votes",
             );
             execute_and_commit_timings.find_and_send_votes_us = find_and_send_votes_time.as_us();
-            commit_time_us
+            (commit_time_us, commit_transaction_statuses)
         } else {
-            0
+            (0, vec![])
```
Comment (carllin) [inline core/src/banking_stage.rs:1300]: Would returning an empty vector is a divergence from current behavior if
1. `executed_transactions_count` is `0`
2. `execution_results` is not empty right

It seems in this situation the current code we would return a `transactions_execute_and_record_status` constructed by

```
 let mut transactions_execute_and_record_status: Vec<bool> = execution_results
            .iter()
            .map(|execution_result| execution_result.was_executed())
            .collect();
```

which would be a vector of all false's with length `execution_results.len()`.

File: core/src/banking_stage.rs [hunk 3]
```diff
@@ -1286,6 +1260,12 @@ impl BankingStage {
             let commit_time_us = commit_time.as_us();
             execute_and_commit_timings.commit_us = commit_time_us;
 
+            let commit_transaction_statuses = tx_results
+                .execution_results
+                .iter()
+                .map(|execution_result| execution_result.was_executed())
+                .collect();
+
```
Comment (taozhu-chicago) [inline core/src/banking_stage.rs:1268]: This moves the commit_transaction_statuses from after `record` to after `bank.commit`, not 100% sure the difference between that two stages, but probably makes more sense to undo cost tracker if transactions are not committed to bank.

File: core/src/banking_stage.rs (hunk 1 above)
Comment (jstarry) [reply core/src/banking_stage.rs:1216]: Yeah we could but I wanted to preserve the behavior of the counter where it only counts transactions that need to be retried due to record failure

File: core/src/banking_stage.rs (hunk 3 above)
Comment (jstarry) [reply core/src/banking_stage.rs:1268]: If record fails, `commit_transactions_result` will be an error and qos will remove all costs

File: core/src/banking_stage.rs (hunk 2 above)
Comment (jstarry) [reply core/src/banking_stage.rs]: Ah yeah, great catch. Fixed

**Review by t-nelson — APPROVED** [review]
we like the red lines

File: core/src/banking_stage.rs (hunk 2 above)
Comment (carllin) [reply core/src/banking_stage.rs]: yeah this bit of code is so fragile, would be good to add some unit tests to cover the expected return values in some of these paths

File: core/src/banking_stage.rs (hunk 1 above)
Comment (carllin) [reply core/src/banking_stage.rs:1216]: `executed_transactions_count` is equivalent to this right: https://github.com/solana-labs/solana/blob/67d3210ebd9055e78ef217c7c6e61559914731f1/core/src/banking_stage.rs#L1222, i.e. the number of transactions that were executed

Comment (jstarry) [reply core/src/banking_stage.rs:1216]: Ah yeah, you're right!

File: core/src/banking_stage.rs (hunk 2 above)
Comment (jstarry) [reply core/src/banking_stage.rs]: Cool, added a test for that path and the debug assert should cover the rest through the existing tests
